feat(workflow_engine): Setup the Task for process_workflow_updates

saponifi3d opened the pull request:
## Description
- Create a celery task namespace for the workflow engine tasks
- Create a task for processing an activity in the workflow_engine -- this will allow us to create a celery task to async execute process workflows when the issue platform creates an activity

In the code I mention there being a follow-up PR: https://github.com/getsentry/sentry/pull/93580 is the initial PR to support the `Activity` / `Group` models.

saponifi3d committed:
- Create a task namespace for workflow engine
- Create a task for handling activity updates

saponifi3d committed:
Start setting up the task for the activity handler

saponifi3d committed:
add the detector_id from the other stuff to the handler so we can see if it's working

saponifi3d committed:
update the task after changing the base branch

saponifi3d committed:
add some tests and make sure things are working

saponifi3d commented on src/sentry/workflow_engine/tasks.py:
```diff
@@ -0,0 +1,73 @@
+    name="sentry.workflow_engine.processors.process_workflow_task",
+    queue="process_workflows",
+    default_retry_delay=5,
+    max_retries=5,
+    soft_time_limit=50,
+    time_limit=60,
+    silo_mode=SiloMode.REGION,
+    taskworker_config=config.TaskworkerConfig(
+        namespace=namespaces.workflow_engine_tasks,
+        processing_deadline_duration=60,
+        retry=retry.Retry(
+            times=3,
+            delay=5,
```
Anyone have any thoughts on what these values for retry / limits should be? just using what we have been.

markstory replied:
There isn't any right/wrong answer here. However, if you are going to define `Retry` you should also define on which exceptions that retrys should be performed. Without that retries will only be automatically performed for `TimeoutError` or an explicit `RetryError`.

kcons replied:
when would we _not_ want to go with `acks_late=True`  in cases where we'd rather double process than drop one?

saponifi3d commented on src/sentry/taskworker/namespaces.py:
```diff
@@ -131,6 +131,10 @@
 uptime_tasks = taskregistry.create_namespace("uptime", app_feature="crons")
 
+workflow_engine_tasks = taskregistry.create_namespace(
+    "workflow_engine", app_feature="workflow_engine"
```
🤔 Should `app_feature` be `shared` here instead?

markstory replied:
The `app_feature` feeds into our cogs tracking system. If you want to see how much your tasks cost compared to other parts of the product you can use `workflow_engine`. Using `shared` will lump your worker costs in with the other shared tasks.

markstory commented on src/sentry/workflow_engine/tasks.py:
```diff
@@ -0,0 +1,73 @@
+
+    The task will get the Activity from the database, create a WorkflowEventData object,
+    and then process the data in `process_workflows`.
+    """
```
Any ideas on what kind of throughput tasks in the workflow_tasks namespace will have?

markstory replied:
We can add this to the alerts taskworker pools when we get US deployed. In other regions throughput will be small enough that it isn't a concern.

saponifi3d committed:
cleanup code a little and fix first broken test

saponifi3d committed:
Address the remaining PR comments and fix remaining failing test
- Add queue to server config
- improved task name / queue
- added acks_late=True

kcons commented on src/sentry/workflow_engine/tasks.py:
```diff
@@ -0,0 +1,74 @@
+    """
+    Process a workflow task identified by the given Activity ID and Detector ID.
+
+    This task will retry up to 3 times with a delay of 5 seconds between attempts.
```
I think we can leave this to the annotation to specify and remove this.

saponifi3d committed:
that time i put def instead of class.

saponifi3d committed:
don't let ai drive alone
